Add JSON schemas for JSON input used in various components

avilcheslopez opened the pull request:
Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

- Relates to #2227
- Fixes #2227, #2296 

Add JSON schemas for JSON input used in various parts of GEOPM. Considering all components that take JSON input perform proper validation (of schema and data), for now it's not urgent to update the code to use a schema validation library and it's sufficient to document the schemas to aid users and contributors/developers.

Ultimately, we will evolve the code to perform more formal validation against the schemas, so this is the first step in that direction.

avilcheslopez committed:
Add JSON schemas for JSON input used in various components

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez commented:
Note: created as a draft for now because I have not decided yet how/where to document the schemas. So for now, I just wanted to publish the schemas to get them reviewed and in the meantime I'll figure out how/where to document them (open to discussion/suggestions).

bgeltz commented on service/docs/source/active_session.schema.json:
```diff
@@ -0,0 +1,35 @@
+    "type": "object",
+    "properties": {
+      "client_pid": {
+        "type": "number"
```
```suggestion
        "type": "integer"
```
The PID will always be an integer.  I don't think we need to support anything else here.
https://json-schema.org/understanding-json-schema/reference/numeric.html

bgeltz replied:
I see that this was done just to match what's already in system_files.py for the ActiveSessions schema.  I'm wondering if we should change this there as well.

avilcheslopez replied:
> I see that this was done just to match what's already in system_files.py for the ActiveSessions schema. I'm wondering if we should change this there as well.

Yeah, I noticed that and opted to just follow what was there for now. But this may be a good opportunity to fix it in the code, then.

bgeltz commented on service/docs/source/active_session.schema.json:
```diff
@@ -0,0 +1,35 @@
+          "type": "string"
+        }
+      },
+      "watch_id": {
+        "type": "number"
+      },
+      "batch_server": {
+        "type": "number"
+      },
```
Same comment here as with the `client_pid`. I think we can change these all to `integer`.

avilcheslopez replied:
> Same comment here as with the `client_pid`. I think we can change these all to `integer`.

Will do!

bgeltz commented on service/docs/source/geopmagent_policy.schema.json:
```diff
@@ -0,0 +1,9 @@
+	"$schema": "http://json-schema.org/draft-04/schema#",
+    "id": "https://geopm.github.io/geopmagent_policy.schema.json",
+    "title": "GEOPM agent policy",
+    "type": "object",
```
How does this handle differences in the way the policy is specified based on the selected Agent?

avilcheslopez replied:
> How does this handle differences in the way the policy is specified based on the selected Agent?

This won't offer deeper name checking (i.e. it won't allow validating that the policy parameter names are correct for a particular agent). This simply specifies that the JSON data must be a set of key-value pairs, where the values can be strings or integers. This is consistent with what the code validates (i.e. FilePolicy is not checking the names).

This is reasonable for a general policy schema. Otherwise, we would have to have a policy schema for each agent (which we could do, I guess).

bgeltz commented on service/docs/source/msrs.schema.json:
```diff
@@ -0,0 +1,70 @@
+    "properties": {
+      "msrs": {
+        "type": "object",
+        "additionalProperties": {
+          "$ref": "#/definitions/msrObject"
```
What purpose do these additionalProperties serve?

avilcheslopez replied:
`additionalProperties` is used when defining the schema for JSON properties of type `object`. In general, because objects are simply key-value pairs, any set of key-value pairs will validate for a property of that type. You can define specific properties with the `properties` keyword, but this on its own won't prevent you from providing other key-value pairs for an object. With the `additionalProperties` keyword, you can specify whether you would like it to accept additional properties besides the ones you've specified. Furthermore, if you do want it to accept additional properties, you can use that keyword to specify the schema for them.

bgeltz reviewed (changes requested):
If you add the license header to these files, it should satisfy CI.

avilcheslopez committed:
Document JSON schemas

Reference the JSON schemas in the documentation, where appropriate for
code that deals with JSON input.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Minor fixes in ActiveSessions JSON schema

Changed the client_pid, watch_id and batch_server type to integer
(from number).

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez commented:
> If you add the license header to these files, it should satisfy CI.

Hmmm... JSON doesn't have a syntax for comments. Looking at other JSON files in the repo, I see they don't have a license header. How did you guys get around this issue for those files?

avilcheslopez committed:
Rename JSON schema file for ActiveSessions

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add documentation for ActiveSessions JSON schema

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

bgeltz commented:
> Hmmm... JSON doesn't have a syntax for comments. Looking at other JSON files in the repo, I see they don't have a license header. How did you guys get around this issue for those files?

Add them to:
https://github.com/geopm/geopm/blob/dev/copying_headers/MANIFEST.EXEMPT

That file specifies which files should be exempt from the test that verifies the license header presence.

avilcheslopez committed:
Move JSON schemas to their own folders, add them to MANIFEST.EXEMPT

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Inline ActiveSessions schema via configure, misc. fixes

- Removed carriage return from JSON schema files
- Fixed minor syntax issue in active_sessions.schema.json

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

avilcheslopez committed:
Add JSON schemas to dist, remove system_fies.py

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>

cmcantalupo commented on service/geopmdpy/system_files.py.in:
```diff
@@ -273,6 +273,15 @@ class ActiveSessions(object):
 
+    The session files are stored in JSON format and follow this schema:
+
+    .. literalinclude:: ../../json_schemas/active_sessions.schema.json
```
This will break when the file is installed rather than being part of the git repository.  This can be a fstring instead.

cmcantalupo commented on service/Makefile.am:
```diff
@@ -75,8 +75,12 @@ EXTRA_DIST = .gitignore \
              geopmdpy/session.py \
              geopmdpy/topo.py \
              geopmdpy/system_files.py \
```
system_files.py should not be in EXTRA_DIST since system_files.py.in is in EXTRA_DIST.  This is because system_files.py is generated by running configure.

cmcantalupo reviewed (changes requested):
Please fix issue with literalinclude of non-installed file inside of an installed file.

avilcheslopez committed:
Inline ActiveSessions schema in Python docstring

Avoid the use of literalinclude as the path is no longer valid once
geopmdpy is installed.

Signed-off-by: Alejandro Vilches <alejandro.vilches@intel.com>
